index.js

```diff
@@ -1,31 +1,109 @@
+  return fromEntries(Object.keys(rules).map((key) => [`react/${key}`, 2]));
+}
+
+const activeRules = filterRules(allRules, (rule) => !rule.meta.deprecated);
+const activeRulesConfig = configureAsError(activeRules);
+
+const deprecatedRules = filterRules(allRules, (rule) => rule.meta.deprecated);
+
 // for legacy config system
 const plugins = [
   'react',
 ];
 
-module.exports = {
-  deprecatedRules: configAll.plugins.react.deprecatedRules,
+const plugin = {
+  deprecatedRules,
   rules: allRules,
   configs: {
-    recommended: Object.assign({}, configRecommended, {
-      parserOptions: configRecommended.languageOptions.parserOptions,
+    recommended: {
       plugins,
-    }),
-    all: Object.assign({}, configAll, {
-      parserOptions: configAll.languageOptions.parserOptions,
+      parserOptions: {
+        ecmaFeatures: {
+          jsx: true,
+        },
+      },
+      rules: {
+        'react/display-name': 2,
+        'react/jsx-key': 2,
+        'react/jsx-no-comment-textnodes': 2,
+        'react/jsx-no-duplicate-props': 2,
+        'react/jsx-no-target-blank': 2,
+        'react/jsx-no-undef': 2,
+        'react/jsx-uses-react': 2,
+        'react/jsx-uses-vars': 2,
+        'react/no-children-prop': 2,
+        'react/no-danger-with-children': 2,
+        'react/no-deprecated': 2,
+        'react/no-direct-mutation-state': 2,
+        'react/no-find-dom-node': 2,
+        'react/no-is-mounted': 2,
+        'react/no-render-return-value': 2,
+        'react/no-string-refs': 2,
+        'react/no-unescaped-entities': 2,
+        'react/no-unknown-property': 2,
+        'react/no-unsafe': 0,
+        'react/prop-types': 2,
+        'react/react-in-jsx-scope': 2,
+        'react/require-render-return': 2,
+      },
+    },
+    all: {
       plugins,
-    }),
-    'jsx-runtime': Object.assign({}, configRuntime, {
-      parserOptions: configRuntime.languageOptions.parserOptions,
+      parserOptions: {
+        ecmaFeatures: {
+          jsx: true,
+        },
+      },
+      rules: activeRulesConfig,
+    },
+    'jsx-runtime': {
       plugins,
-    }),
+      parserOptions: {
+        ecmaFeatures: {
+          jsx: true,
+        },
+        jsxPragma: null, // for @typescript/eslint-parser
+      },
+      rules: {
+        'react/react-in-jsx-scope': 0,
+        'react/jsx-uses-react': 0,
+      },
+    },
   },
 };
+
+plugin.configs['flat/recommended'] = {
+  plugins: { react: plugin },
+  rules: plugin.configs.recommended.rules,
+  languageOptions: { parserOptions: plugin.configs.recommended.parserOptions },
+};
+
+plugin.configs['flat/all'] = {
+  plugins: { react: plugin },
+  rules: plugin.configs.all.rules,
+  languageOptions: { parserOptions: plugin.configs.all.parserOptions },
+};
+
+plugin.configs['flat/jsx-runtime'] = {
+  plugins: { react: plugin },
+  rules: plugin.configs['jsx-runtime'].rules,
+  languageOptions: { parserOptions: plugin.configs['jsx-runtime'].parserOptions },
+};
```

Comment:
this might be more ergonomic if it's a "flat" object with 3 properties, "recommended", "all", "jsx-runtime"?

Reply:
Unlike eslintrc, there are no restrictions on how flat configs can be exported from plugins. However, the convention ESLint recommends is to still export configs as properties of `configs` (https://github.com/eslint/eslint/issues/18095).

Reply:
Sounds good to me at the moment.

In some future major version of this plugin, it would be good to switch to:

 `{ configs: { 'legacy/recommended' /* eslintrc */, 'recommended' /* flat */, ... } }`

Reply:
certainly, if indeed the ecosystem prefers flat config then there will come a time when it doesn't make sense to support anything prior to that.

Reply:
Here's a commit that updates this: https://github.com/mdjermanovic/eslint-plugin-react/commit/bbdaaa7f50303a9a8febf81f03f801af84c3f82e

Reply:
Anything else I could to help finish this?